src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs

```diff
@@ -193,82 +204,48 @@ private ITypeActivatorCache TypeActivatorCache
         /// </remarks>
         public void StartTagHelperWritingScope()
         {
-            StartTagHelperWritingScope(new StringCollectionTextWriter(Output.Encoding));
+            StartTagHelperWritingScope(new HtmlContentBuilder());
         }
 
         /// <summary>
-        /// Starts a new writing scope with the given <paramref name="writer"/>.
+        /// Starts a new writing scope with the given <paramref name="contentBuilder"/>.
         /// </summary>
         /// <remarks>
         /// All writes to the <see cref="Output"/> or <see cref="ViewContext.Writer"/> after calling this method will
         /// be buffered until <see cref="EndTagHelperWritingScope"/> is called.
         /// </remarks>
-        public void StartTagHelperWritingScope(TextWriter writer)
+        public void StartTagHelperWritingScope(IHtmlContentBuilder contentBuilder)
         {
-            if (writer == null)
+            if (contentBuilder == null)
             {
-                throw new ArgumentNullException(nameof(writer));
-            }
-
-            // If there isn't a base writer take the ViewContext.Writer
-            if (_originalWriter == null)
-            {
-                _originalWriter = ViewContext.Writer;
+                throw new ArgumentNullException(nameof(contentBuilder));
             }
 
             // We need to replace the ViewContext's Writer to ensure that all content (including content written
             // from HTML helpers) is redirected.
-            ViewContext.Writer = writer;
 
-            _writerScopes.Push(ViewContext.Writer);
+            _contentBuilderScopes.Push(contentBuilder);
+            RazorTextWriter.Content = contentBuilder;
         }
 
         /// <summary>
         /// Ends the current writing scope that was started by calling <see cref="StartTagHelperWritingScope"/>.
         /// </summary>
-        /// <returns>The <see cref="TextWriter"/> that contains the content written to the <see cref="Output"/> or
+        /// <returns>The <see cref="IHtmlContentBuilder"/> that contains the content written to the <see cref="Output"/> or
```

Comment:
Yeah ignore the docs for now. I just did a find and replace so it's quite messed up.